rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -3020,10 +3046,79 @@ extern "C" rmw_ret_t rmw_wait(
 ///////////                                                                   ///////////
 /////////////////////////////////////////////////////////////////////////////////////////
 
+static rmw_ret_t get_matched_endpoints(
+  dds_entity_t h, dds_return_t (* fn)(
+    dds_entity_t h,
+    dds_instance_handle_t * xs, size_t nxs), std::vector<dds_instance_handle_t> & res)
+{
+  dds_return_t ret;
+  if ((ret = fn(h, res.data(), res.size())) < 0) {
+    return RMW_RET_ERROR;
+  }
+  while ((size_t) ret >= res.size()) {
+    // 128 is a completely arbitrary margin to reduce the risk of having to retry
+    // when matches are create/deleted in parallel
+    res.resize((size_t) ret + 128);
+    if ((ret = fn(h, res.data(), res.size())) < 0) {
+      return RMW_RET_ERROR;
+    }
+  }
+  res.resize((size_t) ret);
+  return RMW_RET_OK;
+}
+
+static void free_builtintopic_endpoint(dds_builtintopic_endpoint_t * e)
+{
+  dds_delete_qos(e->qos);
+  dds_free(e->topic_name);
+  dds_free(e->type_name);
+  dds_free(e);
+}
+
+static std::unique_ptr<dds_builtintopic_endpoint_t,
+  std::function<void(dds_builtintopic_endpoint_t *)>> get_matched_subscription_data(
+  dds_entity_t writer, dds_instance_handle_t readerih)
+{
+  std::unique_ptr<dds_builtintopic_endpoint_t, std::function<void(dds_builtintopic_endpoint_t *)>>
+  ep(dds_get_matched_subscription_data(writer, readerih), &free_builtintopic_endpoint);
+  return ep;
+}
+
+static std::unique_ptr<dds_builtintopic_endpoint_t,
+  std::function<void(dds_builtintopic_endpoint_t *)>> get_matched_publication_data(
+  dds_entity_t reader, dds_instance_handle_t writerih)
+{
+  std::unique_ptr<dds_builtintopic_endpoint_t, std::function<void(dds_builtintopic_endpoint_t *)>>
+  ep(dds_get_matched_publication_data(reader, writerih), &free_builtintopic_endpoint);
+  return ep;
+}
+
+static const std::string request_id_writer_guid_to_string(const rmw_request_id_t & id)
+{
+  std::ostringstream os;
+  os << std::hex;
+  os << std::setw(2) << static_cast<int>(static_cast<uint8_t>(id.writer_guid[0]));
+  for (size_t i = 1; i < sizeof(id.writer_guid) - 1; i++) {
```

Comment:
should be `< sizeof(id.writer_guid)`?
it seems that the last byte won't be copied if not.

Reply:
Yes, absolutely. (I changed it from 0 .. n-1 to 1..n, except I didn't change the upper bound.) There are not enough services/clients in tests to catch this one. This would have been a source of nasty bugs ...